File: src/main/java/xyz/chunkstories/api/entity/traits/serializable/TraitFlyingMode.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.entity.traits.serializable;

import xyz.chunkstories.api.client.LocalPlayer;
import xyz.chunkstories.api.entity.Entity;
import xyz.chunkstories.api.entity.traits.TraitCollidable;
import xyz.chunkstories.api.entity.traits.generic.TraitSerializableBoolean;

/** Keeps track of the flying flag and performs the flying movement logic */
public class TraitFlyingMode extends TraitSerializableBoolean {
	public TraitFlyingMode(Entity entity) {
		super(entity);
	}

	public float flySpeed = 0.125f;
	public boolean noclip = true;

	public void tick(LocalPlayer controller) {
		if (!controller.hasFocus())
			return;

		// Flying resets the entity velocity, if it has one
		getEntity().traits.with(TraitVelocity.class, ev -> ev.setVelocity(0, 0, 0));

		TraitRotation entityRotation = getEntity().traits.get(TraitRotation.class);
		if (entityRotation == null)
			return; // you must be able to rotate to fly

		TraitCollidable entityCollisions = getEntity().traits.get(TraitCollidable.class);
		boolean ignoreCollisions = entityCollisions == null | this.noclip;

		float cameraSpeed = flySpeed;
		if (controller.getInputsManager().getInputByName("flyReallyFast").isPressed())
			cameraSpeed *= 8 * 5f;
		else if (controller.getInputsManager().getInputByName("flyFast").isPressed())
			cameraSpeed *= 8f;

		if (controller.getInputsManager().getInputByName("back").isPressed()) {
			float horizRotRad = (float) ((entityRotation.getHorizontalRotation() + 180f) / 180f * Math.PI);
			float vertRotRad = (float) ((-entityRotation.getVerticalRotation()) / 180f * Math.PI);
			if (ignoreCollisions)
				getEntity().traitLocation.move(Math.sin(horizRotRad) * cameraSpeed * Math.cos(vertRotRad), Math.sin(vertRotRad) * cameraSpeed,
						Math.cos(horizRotRad) * cameraSpeed * Math.cos(vertRotRad));
			else
				entityCollisions.moveWithCollisionRestrain(Math.sin(horizRotRad) * cameraSpeed * Math.cos(vertRotRad), Math.sin(vertRotRad) * cameraSpeed,
						Math.cos(horizRotRad) * cameraSpeed * Math.cos(vertRotRad));
		}
		if (controller.getInputsManager().getInputByName("forward").isPressed()) {
			float horizRotRad = (float) ((entityRotation.getHorizontalRotation()) / 180f * Math.PI);
			float vertRotRad = (float) ((entityRotation.getVerticalRotation()) / 180f * Math.PI);
			if (ignoreCollisions)
				getEntity().traitLocation.move(Math.sin(horizRotRad) * cameraSpeed * Math.cos(vertRotRad), Math.sin(vertRotRad) * cameraSpeed,
						Math.cos(horizRotRad) * cameraSpeed * Math.cos(vertRotRad));
			else
				entityCollisions.moveWithCollisionRestrain(Math.sin(horizRotRad) * cameraSpeed * Math.cos(vertRotRad), Math.sin(vertRotRad) * cameraSpeed,
						Math.cos(horizRotRad) * cameraSpeed * Math.cos(vertRotRad));
		}
		if (controller.getInputsManager().getInputByName("right").isPressed()) {
			float horizRot = (float) ((entityRotation.getHorizontalRotation() + 90) / 180f * Math.PI);
			if (ignoreCollisions)
				getEntity().traitLocation.move(-Math.sin(horizRot) * cameraSpeed, 0, -Math.cos(horizRot) * cameraSpeed);
			else
				entityCollisions.moveWithCollisionRestrain(-Math.sin(horizRot) * cameraSpeed, 0, -Math.cos(horizRot) * cameraSpeed);
		}
		if (controller.getInputsManager().getInputByName("left").isPressed()) {
			float horizRot = (float) ((entityRotation.getHorizontalRotation() - 90) / 180f * Math.PI);
			if (ignoreCollisions)
				getEntity().traitLocation.move(-Math.sin(horizRot) * cameraSpeed, 0, -Math.cos(horizRot) * cameraSpeed);
			else
				entityCollisions.moveWithCollisionRestrain(-Math.sin(horizRot) * cameraSpeed, 0, -Math.cos(horizRot) * cameraSpeed);
		}
	}
}

File: src/main/java/xyz/chunkstories/api/entity/traits/serializable/TraitFlyingMode.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.entity.traits.serializable

import xyz.chunkstories.api.content.json.Json
import xyz.chunkstories.api.content.json.asBoolean
import xyz.chunkstories.api.content.json.asDict
import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.Subscriber
import xyz.chunkstories.api.entity.traits.Trait
import xyz.chunkstories.api.net.Interlocutor
import xyz.chunkstories.api.world.WorldMaster
import java.io.DataInputStream
import java.io.DataOutputStream

/** Keeps track of the flying flag, movement logic has moved to PlayerMovementController in core */
class TraitFlyingMode(entity: Entity) : Trait(entity), TraitSerializable, TraitNetworked<TraitFlyingMode.FlyModeUpdate> {
    override val serializedTraitName = "flyMode"

    var isAllowed: Boolean = false
        set(value) {
            field = value
            if (entity.world is WorldMaster)
                sendMessageController(FlyModeUpdate.SetAllow(field))
        }
    var isFlying: Boolean = false
        set(value) {
            field = value
            if (entity.world is WorldMaster)
                sendMessageController(FlyModeUpdate.SetFlying(field))
            else
                sendMessageAllSubscribers(FlyModeUpdate.SetFlying(field))
        }

    sealed class FlyModeUpdate : TraitMessage() {
        data class SetAllow(val value: Boolean) : FlyModeUpdate() {
            override fun write(dos: DataOutputStream) {
                dos.write(if (value) 1 else 0)
            }
        }

        data class SetFlying(val value: Boolean) : FlyModeUpdate() {
            override fun write(dos: DataOutputStream) {
                dos.write(if (value) 3 else 2)
            }
        }
    }

    override fun readMessage(dis: DataInputStream): FlyModeUpdate {
        return when (dis.read()) {
            0 -> FlyModeUpdate.SetAllow(false)
            1 -> FlyModeUpdate.SetAllow(true)
            2 -> FlyModeUpdate.SetFlying(false)
            3 -> FlyModeUpdate.SetFlying(true)
            else -> throw Exception()
        }
    }

    override fun processMessage(message: FlyModeUpdate, from: Interlocutor) {
        when (message) {
            is FlyModeUpdate.SetAllow -> {
                if (entity.world is WorldMaster) {
                    return // Users cannot set that directly
                }

                isAllowed = message.value
            }
            is FlyModeUpdate.SetFlying -> {
                if (isAllowed) {
                    isFlying = message.value
                }
            }
        }
    }

    override fun whenSubscriberRegisters(subscriber: Subscriber) {
        if(subscriber == entity.traits[TraitControllable::class]?.controller) {
            sendMessageController(FlyModeUpdate.SetAllow(isAllowed))
            sendMessageController(FlyModeUpdate.SetFlying(isFlying))
        }
    }

    override fun serialize() = Json.Dict(mapOf(
                "allow" to Json.Value.Bool(isAllowed),
                "flying" to Json.Value.Bool(isFlying)))

    override fun deserialize(json: Json) {
        val dict = json.asDict ?: return
        isAllowed = dict["allow"].asBoolean ?: false
        isFlying = dict["flying"].asBoolean ?: false
    }
}